In [1]:
import pandas as pd
import numpy as np
import ast
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
%matplotlib inline

In [2]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity,linear_kernel

In [3]:
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader,Dataset,SVD,evaluate

In [14]:
from IPython.display import HTML,Image
pd.set_option('display.max_colwidth', 100)

## Content
1. __Simple Recommender__ :<br>
`1.1 IMDB Top 250`: Top 250 Movies based on calculated IMDB ratings<br> 
`1.2 Recommendation by Genre`: Top Movies for every genre in database
<br>
***
2. __Content Based Recommender__:<br>
`2.1 Movie Description Based:` Recommend movies based on overview and tagline<br>
`2.2 Metadata based`: Recommend movies based on Cast, Director and keywords of movie
***
3. __Collaborative Filtering using Surprise__:<br>
The personalised recommender based on users Past history of ratings and similarity between his and other users rating history
***
4. __Hybrid Recommender__:<br>
Recommeder system based on collaberation of Contend based recommender and collaberative filtering, It leverages are features of both recommender for better recommendations

#### Loadind the data

In [5]:
df = pd.read_csv('movies_metadata.csv')
df.head(2)

/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

     ...     release_date      revenue runtime  \
0    ...       1995-10-30  373554033.0    81.0   
1    ...       1995-12-15  262797249.0   104.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video vote_average  \
0                                        NaN  Toy Story  False          7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False          6.9   

  vote_count  
0     5415.0  
1     2413.0  

[2 rows x 24 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
adult                    45466 non-null object
belongs_to_collection    4494 non-null object
budget                   45466 non-null object
genres                   45466 non-null object
homepage                 7782 non-null object
id                       45466 non-null object
imdb_id                  45449 non-null object
original_language        45455 non-null object
original_title           45466 non-null object
overview                 44512 non-null object
popularity               45461 non-null object
poster_path              45080 non-null object
production_companies     45463 non-null object
production_countries     45463 non-null object
release_date             45379 non-null object
revenue                  45460 non-null float64
runtime                  45203 non-null float64
spoken_languages         45460 non-null object
status                   45379 non-null objec

In [7]:
df = df.drop(['homepage','video'],axis=1)

In [11]:
base_poster_url = 'http://image.tmdb.org/t/p/w185/'
df['poster_path'] = "<img src='" + base_poster_url + df['poster_path'] + "' style='height:100px;'>"

In [8]:
df['genres']  = df['genres'].fillna('[]').apply(ast.literal_eval)
df['genres'] = df['genres'].apply(lambda x:[i['name'] for i in x] if isinstance(x,list) else [])

# 1. Simple Recommender

This is a very basic model based on movies popularity and critical acclaimation, this model does not give personalised recommendation

#####  I will use IMDB's weighted rating formula to construct my chart. Mathematically, it is represented as follows:

##### Weighted Rating (WR) =  ((v/(v+m)).R)+((m/(v+m)).C)
 

##### where,

##### v is the number of votes for the movie
##### m is the minimum votes required to be listed in the chart
##### R is the average rating of the movie
##### C is the mean vote across the whole report

In [9]:
vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
m = vote_counts.quantile(0.9)

In [10]:
df['year'] = pd.to_datetime(df['release_date'],errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [15]:
qualified_df = df[(df['vote_count']>m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())]
qualified_df = qualified_df[['poster_path','title','year','vote_count','vote_average','popularity','genres']]
qualified_df['vote_count'] = qualified_df['vote_count'].astype('int')
qualified_df.head(2)
HTML(qualified_df.head(2).to_html(escape=False))

In [16]:
qualified_df.shape

(4538, 7)

### there are total 4538 movies that are qualified to be considered for the recommender charts.

##### Lets calculated weighted rating according to mentioned IMDB's formulae

In [17]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    
    return (v/(v+m))*R + (m/(v+m)*C)

In [19]:
qualified_df['WR'] = qualified_df.apply(weighted_rating,axis=1)
qualified_df.head(2)
HTML(qualified_df.head(2).to_html(escape=False))

## 1.1 IMDB TOP 250

In [20]:
imdb_top_250 = qualified_df.sort_values('WR',ascending=False).head(250)
imdb_top_250.head(20)
HTML(imdb_top_250.head(15).to_html(escape=False))

## 1.2 Recommendation by genre

In [21]:
def recommendation_by_genre(genre,percentile=0.90):
    
    m = vote_counts.quantile(percentile)
    
    qualified_seg = df[(df['vote_count'] > m) & df['vote_count'].notnull() & df['vote_average'].notnull()] 
    ## This will select the movies with number of votes required, it mostly depend on 'm' as we increase
    ## value of percentile the selected dataframes shrinks
    
    gen_df = pd.DataFrame(columns=qualified_seg.columns)
    # Empty dataset to which we will append the films which belong to given genre
    
    
    for i in range(0,qualified_seg.shape[0]):
        if genre in qualified_seg.iloc[i]['genres']:
            gen_df = pd.concat([gen_df,pd.DataFrame(qualified_seg.iloc[i:i+1])],axis=0)
            
    gen_df = gen_df[['poster_path','title','year','vote_count','vote_average','popularity','genres','original_language']]
    gen_df['vote_count'] = gen_df['vote_count'].astype('int')
    
    gen_df['WR'] = gen_df.apply(lambda x:(x['vote_count']/(x['vote_count']+m) * x['vote_average']) 
                                + (m/(m+x['vote_count']) * C), axis=1)
    ### It will calculate the rating according to the IMDB's Formulae
    
    
    gen_df = gen_df.sort_values('WR',ascending=False).head(250)
    ### sort the movies according to Weighted Average score(WR)
    
    
    return gen_df

In [23]:
recommendation_by_genre('Mystery',percentile=0.95)   
HTML(recommendation_by_genre('Mystery',percentile=0.95).head(10).to_html(escape=False)   )

#### As we move Percentile value, the movies to qualify for recomendation get selected or dropped, Higher the percentile, higher number of votes required to qualify

In [24]:
recommendation_by_genre('Romance',percentile=0.995)
HTML(recommendation_by_genre('Romance',percentile=0.995).head(10).to_html(escape=False))

#### Well these are quite good romantic movies, but Before Sunset is still my favourite Romantic movie, whatever this list says

### The above recommender is not personalised and it will only generate recommendation by genre, irrespective of persons own interest

# 2. Content Based Recommender

To personalise our recommendations more, I am going to build an engine that computes similarity between movies based on certain metrics and suggests movies that are most similar to a particular movie that a user liked. Since we will be using movie metadata (or content) to build this engine, this also known as Content Based Filtering.

Even though content based recommender is still not personalised recommender, we gonna use it letter for hybrid recommender 

I will build 2 content based Recommender based on:

1.Movie Overview and tagline

2.Movie cast,director, genre,keywords

For upcoming recommenders, we gonna need sparse matrix formed using Cosine similarities/linear similarities, so for computation conveninece we will subset of dataset

In [26]:
links_small = pd.read_csv('links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')
df =df.drop([19730, 29503, 35587])    # These movies release date is wrongly added as id, so it is wise to remove them
df['id'] = df['id'].astype('int')
smd = df[df['id'].isin(links_small)]
smd.shape

(9099, 23)

So we will work with 9099 movies

## 2.1 Movie description based recommender 
#### Based on overview and tagline

In [27]:
smd['tagline'] = smd['tagline'].fillna('')
smd['overview'] = smd['overview'].fillna('')
smd['description']  = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')
smd.head(2)

/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the

adult  \
0  False   
1  False   

                                                                                 belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg',...   
1                                                                                                  NaN   

     budget                        genres    id    imdb_id original_language  \
0  30000000   [Animation, Comedy, Family]   862  tt0114709                en   
1  65000000  [Adventure, Fantasy, Family]  8844  tt0113497                en   

  original_title  \
0      Toy Story   
1        Jumanji   

                                                                                              overview  \
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear o...   
1  When siblings Judy and Peter discover an enchanted board game that opens the door to a magical w...   

  popularity  \
0    21.9469   
1    17.0155   

                                                  ...                                                   \
0                                                 ...                                                    
1                                                 ...                                                    

       revenue runtime  \
0  373554033.0    81.0   
1  262797249.0   104.0   

                                                                    spoken_languages  \
0                                           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso_639_1': 'fr', 'name': 'Français'}]   

     status                                    tagline      title  \
0  Released                                             Toy Story   
1  Released  Roll the dice and unleash the excitement!    Jumanji   

  vote_average vote_count  year  \
0          7.7     5415.0  1995   
1          6.9     2413.0  1995   

                                                                                           description  
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear o...  
1  When siblings Judy and Peter discover an enchanted board game that opens the door to a magical w...  

[2 rows x 24 columns]

In [28]:
tf = TfidfVectorizer(analyzer='word',ngram_range=(1,2),min_df=0.0001,stop_words=['english','french'])
tfmatrix = tf.fit_transform(smd['description'])

In [29]:
tfmatrix.shape

(9099, 285340)

### Cosine Similarity

We Will use cosine similarity to calculate numeric quantity that denote similarity between 2 movies.
More info about cosine similarities . https://scikit-learn.org/stable/modules/metrics.html

In [31]:
cosine_sim = cosine_similarity(tfmatrix,tfmatrix)
cosine_sim.shape

(9099, 9099)

In [32]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index,index=smd['title'])

In [33]:
titles.head()

0                      Toy Story
1                        Jumanji
2               Grumpier Old Men
3              Waiting to Exhale
4    Father of the Bride Part II
Name: title, dtype: object

In [34]:
indices.head()

title
Toy Story                      0
Jumanji                        1
Grumpier Old Men               2
Waiting to Exhale              3
Father of the Bride Part II    4
dtype: int64

In [39]:
def get_recommendation(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key= lambda x:x[1], reverse=True)
    sim_scores = sim_scores[1:11]
    movies_indices = [i[0] for i in sim_scores]
    
    return smd.iloc[movies_indices][['poster_path','title','year','description']]

In [41]:
get_recommendation('The Dark Knight')
HTML(get_recommendation('The Dark Knight').to_html(escape=False))

As we can see the recommeder is recommending the movie whose plotline is similar to 'The Dark Knight', which is no surprise that it is showing most of the batman movies

In [43]:
get_recommendation('Before Sunset')
HTML(get_recommendation('Before Sunset').to_html(escape=False))

In [47]:
get_recommendation('Avatar')
HTML(get_recommendation('Avatar').to_html(escape=False))

In [49]:
get_recommendation('Nixon')
HTML(get_recommendation('Nixon').to_html(escape=False))

## 2.2 Metadata Based Recommender

It will be based on cast, crew, genre, keywords

In [50]:
credits = pd.read_csv('credits.csv')
keywords = pd.read_csv('keywords.csv')

In [51]:
keywords.head(2)

id  \
0   862   
1  8844   

                                                                                              keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 10941, 'name': 'disappearance'}, {'id': 15101, 'nam...

In [52]:
credits.head(2)

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender'...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', 'credit_id': '52fe44bfc3a36847f80a7c73', 'gender': ...   

                                                                                                  crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'department': 'Directing', 'gender': 2, 'id': 7879, '...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'department': 'Production', 'gender': 2, 'id': 511, '...   

     id  
0   862  
1  8844

In [53]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] =credits['id'].astype('int')
df['id'] = df['id'].astype('int')
df = df.merge(credits,on='id')
df = df.merge(keywords,on='id')

In [93]:
smd =df[df['id'].isin(links_small)]

In [94]:
smd.shape

(9219, 26)

In [95]:
smd.head(2)         # Check the cast and crew column

adult  \
0  False   
1  False   

                                                                                 belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg',...   
1                                                                                                  NaN   

     budget                        genres    id    imdb_id original_language  \
0  30000000   [Animation, Comedy, Family]   862  tt0114709                en   
1  65000000  [Adventure, Fantasy, Family]  8844  tt0113497                en   

  original_title  \
0      Toy Story   
1        Jumanji   

                                                                                              overview  \
0  Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear o...   
1  When siblings Judy and Peter discover an enchanted board game that opens the door to a magical w...   

  popularity  \
0    21.9469   
1    17.0155   

                                                  ...                                                   \
0                                                 ...                                                    
1                                                 ...                                                    

                                                                    spoken_languages  \
0                                           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso_639_1': 'fr', 'name': 'Français'}]   

     status                                    tagline      title  \
0  Released                                        NaN  Toy Story   
1  Released  Roll the dice and unleash the excitement!    Jumanji   

   vote_average  vote_count  year  \
0           7.7      5415.0  1995   
1           6.9      2413.0  1995   

                                                                                                  cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender'...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', 'credit_id': '52fe44bfc3a36847f80a7c73', 'gender': ...   

                                                                                                  crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'department': 'Directing', 'gender': 2, 'id': 7879, '...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'department': 'Production', 'gender': 2, 'id': 511, '...   

                                                                                              keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 10941, 'name': 'disappearance'}, {'id': 15101, 'nam...  

[2 rows x 26 columns]

In [96]:
smd['cast'] = smd['cast'].apply(ast.literal_eval)
smd['crew'] = smd['crew'].apply(ast.literal_eval)
smd['keywords'] = smd['keywords'].apply(ast.literal_eval)

/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the

In Cast column, there are total 10-15 actors are present for each movie, but we need only top 3-4 main actors who have worked in movie

In [97]:
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x])
smd['cast'] = smd['cast'].apply(lambda x: x[:4] if len(x)>3 else x) # only top 4 actors

/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Now we will get director

In [98]:
def get_director(crew):
    
    for i in crew:
        
        if i['job'] =='Director':
            return i['name']
        
    else:
        return np.nan

In [99]:
smd['director'] = smd['crew'].apply(get_director)
smd = smd.drop('crew',axis=1)

/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [100]:
smd['keywords'] = smd['keywords'].apply(lambda x:[i['name'] for i in x] if isinstance(x,list) else [])

#### Now in cast column there actors such as brad pitt and brad renfro, after we create a soup the computer will consider both brad same  so we need to join first name and last name together

In [101]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(' ','')) for i in x])
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(' ','')))
smd['director'] = smd['director'].apply(lambda x: [x,x])  # We are adding directors name 3 times to give more weightage

### Now using Snowball stemmer we will stem keywords 

In [102]:
stemmer = SnowballStemmer('english')

In [103]:
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])

In [104]:
smd['keywords'] = smd['keywords'].apply(lambda x:[str.lower(i.replace(' ','')) for i in x])

In [105]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']

In [106]:
smd['soup'] = smd['soup'].apply(lambda x:' '.join(x))

In [107]:
count = CountVectorizer(analyzer='word',ngram_range=(1,2),min_df=0,stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [108]:
count_matrix.shape

(9219, 132041)

In [109]:
cosine_sim = cosine_similarity(count_matrix,count_matrix)

In [110]:
smd =  smd.reset_index()
titles = smd['title']
indices = pd.Series(titles.index, index = smd['title'])

In [111]:
def get_recommendation_2(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key= lambda x:x[1], reverse=True)
    sim_scores = sim_scores[1:11]
    movies_indices = [i[0] for i in sim_scores]
    j= smd.iloc[movies_indices][['poster_path','title','year','cast','director']]
    j['director'] = j['director'].apply(lambda x:x[0])
    return j

As Our Cosine function is changed, lets see what recommendation we get now

In [112]:
get_recommendation_2('The Dark Knight')
HTML(get_recommendation_2('The Dark Knight').to_html(escape=False))

Now Most of the movies recommended by recommender  for dark knight are dominated by christian bale, and christopher nolan

In [113]:
get_recommendation_2('Avatar')
HTML(get_recommendation_2('Avatar').to_html(escape=False))

In [114]:
get_recommendation_2('Before Sunset')
HTML(get_recommendation_2('Before Sunset').to_html(escape=False))

# 3. Collaborative Filtering

The above recommendation models has one limitation, It is not user centric, It will only recommend the movies which are close to the given movies.
I will use Surprise library for more user centric recommendations

In [115]:
reader = Reader()

In [116]:
ratings = pd.read_csv('ratings_small.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [117]:
ratings['movieId'].nunique()

9066

In [118]:
data = Dataset.load_from_df(ratings[['userId','movieId','rating']],reader)
data.split(n_folds=5)
svd = SVD()
evaluate(algo=svd,data=data,measures=['RMSE','MAE'])

/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/surprise/evaluate.py:66: UserWarning: The evaluate() method is deprecated. Please use model_selection.cross_validate() instead.
  'model_selection.cross_validate() instead.', UserWarning)
/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/surprise/dataset.py:193: UserWarning: Using data.split() or using load_from_folds() without using a CV iterator is now deprecated. 
  UserWarning)


Evaluating RMSE, MAE of algorithm SVD.

------------
Fold 1
RMSE: 0.8976
MAE:  0.6897
------------
Fold 2
RMSE: 0.8915
MAE:  0.6879
------------
Fold 3
RMSE: 0.8993
MAE:  0.6939
------------
Fold 4
RMSE: 0.9022
MAE:  0.6942
------------
Fold 5
RMSE: 0.8959
MAE:  0.6900
------------
------------
Mean RMSE: 0.8973
Mean MAE : 0.6911
------------
------------


CaseInsensitiveDefaultDict(list,
                           {'rmse': [0.8976340079189299,
                             0.8914695394582584,
                             0.8993006431621312,
                             0.9021528138098897,
                             0.895931173341792],
                            'mae': [0.689713853205912,
                             0.6878668373981629,
                             0.6939102476430006,
                             0.6942286979321564,
                             0.6900248249026474]})

In [119]:
trainset = data.build_full_trainset()
svd.train(trainset)

/Users/suyashchavan/anaconda3/lib/python3.7/site-packages/surprise/prediction_algorithms/algo_base.py:51: UserWarning: train() is deprecated. Use fit() instead
  warnings.warn('train() is deprecated. Use fit() instead', UserWarning)


In [122]:
ratings[ratings['userId']==10][:10]      # These are first 10 ratings given by user with userid10

userId  movieId  rating  timestamp
744      10       50     5.0  942766420
745      10      152     4.0  942766793
746      10      318     4.0  942766515
747      10      344     3.0  942766603
748      10      345     4.0  942766603
749      10      592     3.0  942767328
750      10      735     4.0  942766974
751      10     1036     3.0  942767258
752      10     1089     3.0  942766420
753      10     1101     2.0  942767328

In [124]:
svd.predict(uid=10,iid=33)

Prediction(uid=10, iid=33, r_ui=None, est=3.565545318581256, details={'was_impossible': False})

The model predict rating given by userid 10 to movie id 33 is 3.56. This is purely on how other user have given rating to the movie and how their ratings are similar to userid10

# 4. Hybrid Recommender

We will collaborate the content based(Metadata based) and collaborative filtering recommenders

In [125]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [126]:
id_map = pd.read_csv('links_small.csv')
id_map.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [127]:
id_map = id_map[['movieId','tmdbId']]
id_map['movieId'].nunique()

9125

In [128]:
id_map.columns= ['movieId','id']
id_map = id_map.merge(smd[['title','id']],on='id').set_index('title')
id_map.head()

movieId       id
title                                        
Toy Story                          1    862.0
Jumanji                            2   8844.0
Grumpier Old Men                   3  15602.0
Waiting to Exhale                  4  31357.0
Father of the Bride Part II        5  11862.0

In [129]:
indices_map = id_map.set_index('id')
indices_map.head()

movieId
id              
862.0          1
8844.0         2
15602.0        3
31357.0        4
11862.0        5

In [130]:
indices.head(3)

title
Toy Story           0
Jumanji             1
Grumpier Old Men    2
dtype: int64

In [131]:
titles.head(3)

0           Toy Story
1             Jumanji
2    Grumpier Old Men
Name: title, dtype: object

In [132]:
def hybrid_recommender(userId,title):
    idx = indices[title]
    tmdbId = id_map.loc[title]['id']
    movie_id = id_map.loc[title]['movieId']
    
    sim_scores = list(enumerate(cosine_sim[int(idx)]))
    sim_scores = sorted(sim_scores,key=lambda x:x[1],reverse=True)
    sim_scores = sim_scores[1:51]
    movie_indices = [x[0] for x in sim_scores]
    movies = smd.loc[movie_indices][['poster_path','title','vote_count','vote_average','year','id']]
    movies['est'] = movies['id'].apply(lambda x : svd.predict(userId,indices_map.loc[x]['movieId']).est)
    movies = movies.sort_values('est',ascending=False)
    
    return movies.head(10)

In [133]:
hybrid_recommender(1,'Casino')
HTML(hybrid_recommender(1,'Casino').to_html(escape=False))

In [137]:
hybrid_recommender(120,'Casino')
HTML(hybrid_recommender(120,'Casino').to_html(escape=False))

We can clearly see that Recommendation for userId =1 & userID=120 are changed, based on users ratings history(indirectly say his liking)

In [134]:
hybrid_recommender(4,'Mad Max: Fury Road')
HTML(hybrid_recommender(4,'Mad Max: Fury Road').to_html(escape=False))

In [138]:
hybrid_recommender(40,'Mad Max: Fury Road')
HTML(hybrid_recommender(40,'Mad Max: Fury Road').to_html(escape=False))